### ISSUE TITLE
Need CircularIO and BytesIO running at the same time

### ISSUE THREAD

**Issue by gabeale** [labthings/picamerax#15]
Hi,
As far as I have tested Picamerax is really amazing and working very well. The web streaming in particular is awesome.
I’m using it for my home surveillance system that I’m building but I’m stuck with one issue that I don’t know whether it is depending on Picamerax code or on my poor python knowledge!

Basically I would need to run picamera with both CircularIO an BytesIO instances at the same time and independently each other: is it doable?

Scope is to run MJPEG web streaming while having the possibility to record video to a file in case motion is detected.
Motion trigger is done externally on a different machine elaborating the MJPEG streaming and returning to the picamerax the “true” condition for the recording cycle.
Thank you in advance for help you are giving me.
Cheers

**Comment by jtc42** [comment]
Hey! So you're able to process multiple video feeds from the picamera library by using splitter ports. An example (used to record to multiple resolutions) can be found here https://picamera.readthedocs.io/en/release-1.10/recipes2.html#recording-at-multiple-resolutions 

What you can do is use the `splitter_port` argument to record video to two separate outputs, which I *think* is what you want.

If you don't mind, I'll close this issue since it's something that's really being handled upstream by `picamera` rather than specifically `picamerax`. Hopefully this helps though!

**Comment by gabeale** [comment]
Hi!
thanks for your prompt answer. About _splitter_port_ argument I have already tested as described in the documentation link you have reported: this is working smoothly but is not exactly what I'm looking for. Basically I'm looking for something in the middle between Web Streaming (ref. https://picamerax.readthedocs.io/en/latest/recipes2.html#web-streaming) and Recording to a Circular Stream (ref. https://picamerax.readthedocs.io/en/latest/recipes1.html#recording-to-a-circular-stream). Concept is that web streaming is broadcasted on the web where another machine detects the motion (e.g. running _DOODS image processing) and returning to Picamerax the trigger when motion is detected (in replacement of _random.randint(0, 10) == 0_ of above mentioned Circular Stream recipes). Unfortunately I'm not able to concatenate these two recipes because when motion is detected starts the _while loop_ and web streaming is suspended so also motion algorithms won't be updated as well and _while loop_ is perpetual!
On the other hand, if I make two separated Python scripts, it is also not working because they are calling at the same time the same camera (using _io.BytesIO()_ and _PiCameraCircularIO_ registers respectively).
Bottom line: I'm looking for a python script concatenating the above mentioned recipes. 
Thanks for your help! :-)
Cheers!

**Comment by jtc42** [comment]
You'll need to run your loop in another thread. Have a look at the Python `threading` library. The PiCamera can be accessed by multiple threads (but not processes, hence the behaviour you're seeing). Closing this issue now as it's broader than this package. Good luck with the project though!

**Comment by gabeale** [comment]
Dear Joel,
I have tested the Python "threading" library as you have suggested, without success.
Just for my knowledge, according to you, it is possible the web streaming (ref. https://picamerax.readthedocs.io/en/latest/recipes2.html#web-streaming) whilst recording a motion (ref. https://picamerax.readthedocs.io/en/latest/recipes1.html#recording-to-a-circular-stream) with the same Picamera?
Thx
P.S. I'm not going to re-open issue in subject! ;-)

**Comment by jtc42** [comment]
That's surprising... We've done basically exactly this in the past for the OpenFlexure Microscope software, where we have multiple splitter ports recording to multiple different outputs simultaneously.

You should be able to start each splitter port recording at the same time, and assuming you pass different output objects to each, they should be able to be processed simultaneously. 

I'll leave the issue closed for the reasons outlined previously, but one other resource you might find useful is https://gitlab.com/openflexure/openflexure-microscope-server since that's where we implemented similar functionality to what you're describing, using threading and splitter ports.

**Comment by Esser50K** [comment]
One of the options was already presented, you can use the splitter ports. But if you want the exact same stream for streaming and recording you can just have an object that writes to both at the same time. Here is an example from PiCameleon with the output holder: https://github.com/Esser50K/PiCameleon/blob/main/picameleon/outputs/output_holder.py#L47

**Comment by jtc42** [comment]
> One of the options was already presented, you can use the splitter ports. But if you want the exact same stream for streaming and recording you can just have an object that writes to both at the same time. Here is an example from PiCameleon with the output holder: https://github.com/Esser50K/PiCameleon/blob/main/picameleon/outputs/output_holder.py#L47

Great suggestion, thanks!

**Comment by gabeale** [comment]
Excellent guys!
I'll read carefully your post and try to apply it to my project. I'll let you know!
Have a great day! 
Ciao
